### PR TITLE
Basic integration of OpenTelemetry

### PR DESCRIPTION
**Pull request by eranrund**
### Motivation

We'd like to get visibility into where time is spent in the life cycle of a block. The solution proposed here uses the [OpenTelemetry](https://opentelemetry.io/) spec to send trace data to [Jaeger](https://www.jaegertracing.io/).

This allows generation of traces that look like this:
image

The data is also available via a JSON api, so we could build tools around it.

### In this PR

OT was chosen because it seems like a reasonable abstraction for tracing backends, and Jaeger was chosen because I was able to get it to display data grouped by block IDs. There's a discrepancy between what the various tracing frameworks I looked at (Jaeger included) and what we are trying to do. The classic/intended use of a distributed tracing framework is to trace the life cycle of a single request - e.g. an API request comes in, triggers a few requests to related microservices, work is done and eventually a response is returned to the original caller. This constitutes a single `trace`, inside which multiple `spans` represents the various places time was spent. 
In our case, we are trying to get data on a block level and that doesn't really match the model described above. The workaround I came with, that worked well with Jaeger (also tried Zipkin but couldn't get it to display data in a useful way), is to manually override the Trace ID with one that is derived from the block number.

To test this, I ran my own [Jaeger container](https://www.jaegertracing.io/docs/1.28/getting-started/#all-in-one), and injected environment variables inside `setup_default_tracer_with_tags`:
```
            env::set_var("OTEL_EXPORTER_JAEGER_AGENT_HOST", "1.2.3.4");
            env::set_var("OTEL_EXPORTER_JAEGER_AGENT_PORT", "6831");
```

### Future Work
* We need `jaeger-agent` to run alongside each of the containers we want to collect metrics from. Metrics are reported synchronously over UDP, so having this run on the same machine is important for performance. Asynchronous reporting requires committing to an async runtime (tokio or async-std) so I opted out of that for now. I have discussed this need with the ops team and this will likely be a Q1 priority.

### CODE REVIEW

File: fog/test-client/src/test_client.rs
Comment (eranrund) [inline fog/test-client/src/test_client.rs:257]: Nothing changed here other than sticking all the existing code under the span.

File: fog/test-client/src/test_client.rs
Comment (eranrund) [inline fog/test-client/src/test_client.rs:413]: Nothing changed here other than sticking all the existing code under the span.

**Comment by eranrund** [comment]
I just realized this is missing tracing for fog-ledger-server, so I am going to add that.

**Comment by eranrund** [comment]
A trace that includes fog-ledger-server:
image

File: fog/ingest/server/src/worker.rs
Comment (garbageslam) [inline fog/ingest/server/src/worker.rs:156]: It seems that sometimes we're doing
```
let mut span = ...;
...
span.end()
```

And sometimes we do this thing where the code is stuffed in a lambda

Fwiw I find it a little more readable when the span is a separate statement, but I'm not gonna block the PR
Especially in the case where lambda returns something

They are actually equivalent though right? Or does one of them offer extra functionality?

File: fog/test-client/src/test_client.rs
Comment (garbageslam) [inline fog/test-client/src/test_client.rs:258]: For my understanding, why are we making a new Tracer context here? There should be an outer one in test_transfer

File: fog/ingest/server/src/worker.rs
Comment (eranrund) [reply fog/ingest/server/src/worker.rs:156]: The `in_span` does not let you control aspects of the span you are creating like the trace id or start/end times, which is why it is only useful when you're already inside a span that has been initialized with the desired TraceId.

File: fog/test-client/src/test_client.rs
Comment (eranrund) [reply fog/test-client/src/test_client.rs:258]: They all use the same global tracer context, it's easier than passing it around.

**Comment by garbageslam** [comment]
Let's try to merge this before the test clients fixup PR. I think Jason is deploying a custom build of that and it doesn't matter for now if it's actually merged and releases. But we do want to release servers with the tracing functionality I think pretty soon

**Review by garbageslam — APPROVED** [review]

**Review by samdealy — APPROVED** [review]
Looks great. 

One thing I'd request (perhaps in a follow-up PR) is to document how / when to add a span to the trace, and how to access the info on Jaeger.  Doc would answer questions like "What sorts of features are good candidates for adding to the tracing system?". Could live as a README or a google doc somewhere.

**Comment by eranrund** [comment]
@jcape are you interested in reviewing this before I merge?

**Review by jcape — APPROVED** [review]
LGTM

**Pull request merged**
